### DISCUSSION TITLE
Triangle Export

### DISCUSSION THREAD

**Discussion by KiY0On** [casact/chainladder-python#314] [Q&A]
Hi,

I am working on using the Triangle object from the chain-ladder package and seeking some help in using it in my project.

At the current stages:

1. I have the flat data, which I have it as a pandas dataframe as I needed to format before feeding it as the triangle data.
2. I have 2 index - one is the type of amounts such as incurred loss, case outstanding, paid loss, written prem and earned prem; the other is the reserve study group. I am working on the syntax to create these triangles.

I am concerned that once I create the triangles, would I be able to export all of the triangles as a single flat csv file to be used for analysis. I would like to use the CAS repository for the analytical part, but I wanted to try turning the flat data into the triangle and exporting out first. 

I would appreciate any idea or advice on this subject. Thank you!

**Comment by jbogaardt** [comment]
The `Triangle` object has a `to_frame` method. This should allow you to export a multidimensional triangle to a pandas dataframe with all relevant info intact.

```python
>>> import chainladder as cl
>>> print(cl.Triangle.to_frame.__doc__)
 Converts a triangle to a pandas.DataFrame.

        Parameters
        ----------
        origin_as_datetime : bool
            Whether the origin vector should be converted from PeriodIndex
            into a datetime dtype. Default is False.
        keepdims : bool
            If True, the triangle will be converted to a DataFrame with all
            dimensions intact.  The argument will force a consistent DataFrame
            format regardless of whether any dimensions are of length 1.
        implicit_axis : bool
            When keepdims is True, this denotes whether to include the implicit
            valuation axis in addition to the origin and development.
        Returns
        -------
            pandas.DataFrame representation of the Triangle.
```

**Comment by kennethshsu** [comment]
Hi @KiY0On, it looks like you are new. Welcome! There's a starter tutorial which introduces you to most of the basic functionalities of the package [here](https://chainladder-python.readthedocs.io/en/latest/tutorials/triangle-tutorial.html). Did you get a chance to take a look?

**Reply by KiY0On** [reply]
Hi, I took at look at the starter tutorial and it's really helpful. Just a quick question. I am playing around with the different development lags such as monthly or semi annually, but I am having a trouble formatting my data. Is there a way for me to access what the proper formatting is?

**Reply by kennethshsu** [reply]
Do you have sample code? What do you mean by having trouble formatting the data?

It looks like you might want to access the `grain()` method. More details [here](https://chainladder-python.readthedocs.io/en/latest/triangle.html?highlight=grain#triangle-grain).

**Reply by KiY0On** [reply]
I have trouble getting the development data. My data has the development period such as 6, 12, 18 ... so on, but I am not quite sure how to convert this into the development format. I was able to use the example code to get to yearly format tho.

In below, I was trying to calculate the calendar month based on my development period, but it doesn't seem like working properly.

df_input["und_yr"] = pd.to_datetime(df_input["und_yr"],format = '%Y')
df_input['Cal Year'] = (df_input["und_yr"].dt.year + df_input["im"] / 12 - 1).astype(int)
df_input['Cal Month'] = (df_input["und_yr"].dt.month + df_input["im"].mod(12)).astype(int)

To use the grain method for semi annual development, I would use .grain('OYDM') right? Does that mean my development should be in 'YYYY/MM'?

**Reply by kennethshsu** [reply]
Your data needs to have at least semi-annual granularity. To see semi-annual development, your `D` should be an `S` for Semi-Annual or `H` for half-year. 

Try `grain('OYDS')` or `grain('OYDH')`.

**Reply by KiY0On** [reply]
I've tried grain('OYDS'), but it's not outputting the right development for me.

Do you mind explaining the granularity of the data?
For example:
df has calendar year data as 2000-06. 2001-01, 2001-06... and if I use the Triangle(dvelopment = 'Cal year').grain('OYDS') does that spit out the semi-annual development?

Also, for multiple indexed triangles, is there a way to visualize one triangle?

**Reply by kennethshsu** [reply]
Is it possible for you to post a sample data, so I can take a look?

There's no way to visualize multiple indexed triangles. You can try to sum all triangles by doing `sum()`.

**Reply by kennethshsu** [reply]
Also, I think the calendar year data needs to be January and July, not January and June. Can you try to make sure your data is at the right granularity level?

**Reply by KiY0On** [reply]
Yes, I will post the sample data.

In the meantime, just one more quick question. Once I feed the data in to create a multi-indexed data, can I output the multi indexed triangles as a csv file? But output them in the triangle formats.

Edit: I was able to output one triangle in the triangle format. However, with the multi-indexed triangles, I was trying to figure out how to iterate each triangle to output as csv. What should be a syntax to access one triangle out of multi-indexed triangle?

**Reply by kennethshsu** [reply]
For that, you'll need to loop through your indexes and do `.to_frame()` at each index. I don't think there's another way. @jbogaardt?

**Reply by KiY0On** [reply]
I know how to access a column of the dataframe like df['paid']. But to access by using index, what would the syntax look like? df(columns = 'paid', index = '192') ?

**Reply by kennethshsu** [reply]
You can do something like this:

```python
raa = cl.load_sample("raa")
raa[raa.origin == "1983"][raa.development == 36]
```

**Comment by KiY0On** [comment]
[Sample Data.csv](https://github.com/casact/chainladder-python/files/9144571/Sample.Data.csv)
Here is the sample data of one study group.

[NDT EU TFtest.csv](https://github.com/casact/chainladder-python/files/9144633/NDT.EU.TFtest.csv)
Here is the output I got from the initial run.

There must be something wrong with when I create the development column...

Thank you so much for the help and apologies for asking too many questions.

**Reply by kennethshsu** [reply]
Can you also post your python code as well? 

Am I understanding this correctly? I think you used `under_yr` as your origin for underwriting_year, and `Development Period` as your ages? `Development Period` will need to be a date-like value, so instead of age 12, put the same year as origin. Is your `Commission Amount` your values?

See if you can post your python code I can help take a look.

**Reply by KiY0On** [reply]
Yes, I am trying to create `Age factor` by using `under_yr` + `Development Period`, which is in a 6month step in order to create a semi annual development triangle. Sorry, I am not familiar with the data-like value. 
```
df_input["und_yr"] = pd.to_datetime(df_input["und_yr"],format = '%Y')
df_input['Cal Year'] = (df_input["und_yr"].dt.year + df_input["im"] // 12).astype(str)
df_input['Cal Season'] = (df_input["im"].mod(12)).astype(str)
df_input['development'] = (df_input["Cal Year"] +'-'+ df_input["Cal Season"])
df_input['development'] = pd.to_datetime(df_input['development'],format = '%Y-%M')
```

**Reply by kennethshsu** [reply]
Date-like values are what they sound like, the value should look like a date, and not an integer. For example, for `und_yr = 1996` at age `12`: instead of `12`, we want `1996`. 

Consider this code that converts the ages to valuation.

```python
df_input = pd.read_csv("/Users/kennethhsu/Downloads/Sample.Data.csv")

df_input["und_yr"] = pd.to_datetime(df_input["und_yr"], format="%Y")
df_input["Development Period"] = df_input["Development Period"].astype(int)

f = lambda x: x["und_yr"] + pd.DateOffset(months=x["Development Period"])
df_input["valuation"] = df_input.apply(f, axis=1) - pd.DateOffset(days=1)

df_input["und_yr"] = df_input["und_yr"].dt.strftime("%Y")
df_input["valuation"] = df_input["valuation"].dt.strftime("%Y-%m")
df_input
```

Once we do this, we will then be able to port the data frame into the triangle format by doing `cl.Triangle(...)`.

Also, I noticed that the sample data that you posted contains lots of blank rows, you probably removed them on purpose. But on the off chance that it's a mistake, be extra careful when you are working with a data frame that contains blank rows.

**Reply by KiY0On** [reply]
Thank you so much.
